File: 239.sliding-window-maximum.java
import java.util.ArrayDeque;
import java.util.LinkedList;

/*
 * @lc app=leetcode id=239 lang=java
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start
class Solution {
    /**
     * 暴力法
     * 
     * @param nums
     * @param k
     * @return
     */
    public int[] maxSlidingWindow1(int[] nums, int k) {
        int[] maxArray = new int[nums.length - k + 1];
        for (int i = 0; i <= nums.length - k; ++i) {
            int max = Integer.MIN_VALUE;
            for (int j = i; j < i + k; ++j) {
                max = Math.max(max, nums[j]);
            }
            maxArray[i] = max;
        }
        return maxArray;
    }

    /**
     * 双端队列法
     * 
     * @param nums
     * @param k
     * @return
     */
    public int[] maxSlidingWindow2(int[] nums, int k) {
        int[] maxArray = new int[nums.length - k + 1];
        ArrayDeque<Integer> deque = new ArrayDeque<>();
        for (int i = 0; i < nums.length; ++i) {
            if (!deque.isEmpty() && deque.peekLast() < i - k + 1) {
                deque.removeLast();
            }
            while (!deque.isEmpty() && nums[deque.peekFirst()] < nums[i]) {
                deque.removeFirst();
            }
            deque.addFirst(i);
            if (i >= k - 1) {
                maxArray[i - k + 1] = nums[deque.peekLast()];
            }
        }
        return maxArray;
    }

    /**
     * 暴力法优化 最优解
     * 
     * @param nums
     * @param k
     * @return
     */
    public int[] maxSlidingWindow3(int[] nums, int k) {
        int len = nums.length;
        if (len == 0 || k > len) {
            return new int[0];
        }
        // 定义结果数组
        int[] res = new int[len - k + 1];
        // maxInd记录每次最大值的下标，max记录最大值
        int maxInd = -1, max = Integer.MIN_VALUE;

        for (int i = 0; i < len - k + 1; i++) {
            // 判断最大值下标是否在滑动窗口的范围内
            if (maxInd >= i) {
                // 存在就只需要比较最后面的值是否大于上一个窗口最大值
                if (nums[i + k - 1] > max) {
                    max = nums[i + k - 1];
                    // 更新最大值下标
                    maxInd = i + k - 1;
                }
            }
            // 如果不在就重新寻找当前窗口最大值
            else {
                max = nums[i];
                for (int j = i; j < i + k; j++) {
                    if (max < nums[j]) {
                        max = nums[j];
                        maxInd = j;
                    }
                }
            }
            res[i] = max;
        }
        return res;

    }
}
// @lc code=end

File: 239.sliding-window-maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#
# https://leetcode.com/problems/sliding-window-maximum/description/
#
# algorithms
# Hard (41.56%)
# Likes:    2868
# Dislikes: 164
# Total Accepted:    241.7K
# Total Submissions: 581K
# Testcase Example:  '[1,3,-1,-3,5,3,6,7]\n3'
#
# Given an array nums, there is a sliding window of size k which is moving from
# the very left of the array to the very right. You can only see the k numbers
# in the window. Each time the sliding window moves right by one position.
# Return the max sliding window.
# 
# Follow up:
# Could you solve it in linear time?
# 
# Example:
# 
# 
# Input: nums = [1,3,-1,-3,5,3,6,7], and k = 3
# Output: [3,3,5,5,6,7] 
# Explanation: 
# 
# Window position                Max
# ---------------               -----
# [1  3  -1] -3  5  3  6  7       3
# ⁠1 [3  -1  -3] 5  3  6  7       3
# ⁠1  3 [-1  -3  5] 3  6  7       5
# ⁠1  3  -1 [-3  5  3] 6  7       5
# ⁠1  3  -1  -3 [5  3  6] 7       6
# ⁠1  3  -1  -3  5 [3  6  7]      7
# 
# 
# 
# Constraints:
# 
# 
# 1 <= nums.length <= 10^5
# -10^4 <= nums[i] <= 10^4
# 1 <= k <= nums.length
# 
# 
#

# @lc code=start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        deque = []
        def push(n):
            while deque and deque[-1] < n: #删除前n 个比填入数字小的数组 当前只保存最大的值
                deque.pop()
            deque.append(n)
        
        def max():
            return deque[0]

        def remove(n):
            if deque and deque[0] == n: 
                deque.pop(0)

        ans = []
        if not nums or not k:
            return ans
        for i, value in enumerate(nums):
            if i < k - 1: #这个地方先填充k-1个元素
                push(value)
            else:
                push(value) #填充第三个元素时 开始比较
                ans.append(max())
                remove(nums[i - k + 1])
        return ans
        
# @lc code=end

